src/expr/node_traversal.h

```diff
@@ -0,0 +1,148 @@
+/*********************                                                        */
+/*! \file node_traversal.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Iterators for traversing nodes.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__SMT_UTIL__NODE_TRAVERSAL_H
+#define CVC4__SMT_UTIL__NODE_TRAVERSAL_H
+
+#include <cstddef>
+#include <iterator>
+#include <vector>
+
+#include "expr/node.h"
+
+namespace CVC4 {
+
+// Iterator for traversing a node in post-order
+// It does DAG-traversal, so indentical sub-nodes will be visited once only.
+class NodeDfsIterator
+{
+ public:
+  // Construct a traversal iterator beginning at `n`
+  NodeDfsIterator(TNode n, bool postorder);
+  // Construct an end-of-traversal iterator
+  NodeDfsIterator(bool postorder);
+
+  // STL type definitions for an iterator
+  using value_type = TNode;
+  using pointer = TNode*;
+  using reference = TNode&;
+  using iterator_category = std::forward_iterator_tag;
+  using difference_type = std::ptrdiff_t;
```

Comment:
[Typedefs go before constructors.](https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#declaration-order)

Reply:
Done in c394cea1a